<!DOCTYPE html>
<html>
<head>
<title>The Locked Box Exercise</title>
</head>
<body>
    <h1>The Locked Box Exercise</h1>
    <p>Please check the code and the console from JS section. Thank you!</p>
    <script>
        /**
         * EXERCISE FOR ERROR HANDLING
         * 
        */
        //Write a function called withBoxUnlocked that takes a function value as argument, unlocks the box, 
        //runs the function, and then ensures that the box is locked again before returning, regardless of 
        //whether the argument function returned normally or threw an exception.
        const box = {
            locked: true,
            unlock() { this.locked = false; },
            lock() { this.locked = true;  },
            _content: [],
            get content() {
                if (this.locked) throw new Error("Locked!");
                return this._content;
             }};
        function withBoxUnlocked(body) {
            // Your code here.
            if(box.locked == true){
                box.unlock();
            }
            try{
                body();
            }
            finally{
                box.lock();
            }
            }

        withBoxUnlocked(function() {
        box.content.push("gold piece");
        });

        try {
            withBoxUnlocked(function() {
            throw new Error("Pirates on the horizon! Abort!");
        });
        } catch (e) {
            console.log("Error raised: " + e);
        }
        console.log(box.locked);
        // â†’ true
    </script>
</body>
